Add intercept on decorator

idiotsky opened the pull request:
Add support for register service with Interceptor and multiple decorator,  apply Interceptor to the outermost decorator,

Commit:
refactor RegistrationExtensions.cs

Created a new ProxyHelpers class to improve code structure and readability. This class centralizes proxy operations that were previously distributed across multiple classes. Refactored RegistrationExtensions.cs to replace direct method calls with references to the newly created ProxyHelpers.

Commit:
Add service middleware and associated tests for DynamicProxy

DelegateMiddleware class is copied from autofac, after autofac add the overload method, remove this. Additionally, ServiceMiddlewareRegistrationExtensions has been added to provide static methods for registering middleware services. Tests have been provided to ensure correct interception of public interfaces. The Moq testing library has been added as a package reference.

tillig commented:
A description on the PR here, possibly explaining it and linking back to the source issue, would be good. That way in a year from now when we come back and look to see what the point of the PR was (say, during supporting an issue) we have that to reference.

(Update the actual PR description; don't respond with the description here as a comment. Thanks!)

tillig commented:
From what I gather, [this PR needs to go through first](https://github.com/autofac/Autofac/pull/1417) to help this one. I probably won't look at this one until that one is done.

idiotsky commented:
I simply copied the three old RegisterServiceMiddleware<TService> overload methods and made the generic TService a parameter called serviceType. I ran the code coverage locally because this method doesn't have any existing tests or code coverage, and as a result, the related methods also lack code coverage. Could you please add code coverage for this method so I can add tests as well?
`
public static void RegisterServiceMiddleware<TService>(this ContainerBuilder builder, string descriptor, PipelinePhase phase, Action<ResolveRequestContext, Action<ResolveRequestContext>> callback)
    {
        builder.RegisterServiceMiddleware<TService>(descriptor, phase, MiddlewareInsertionMode.EndOfPhase, callback);
    }
`

tillig commented:
I can't lie, while we love a PR coming in, we sort of need to rely on the community to help bolster the existing tests, too. If this is going to require me to add tests so you can just copy them, it's going to be waiting a while. I don't mean that to sound challenging, but I do want to set expectations. There's not a dedicated "Autofac team" paid and cranking along on this. There are like two free-time maintainers, one of whom just switched from .NET to Ruby on Rails so... yeah.

tillig commented:
[Whoops, that comment was for a different PR, sorry!]

idiotsky commented:
The new changes is base on Autofac 8.2.0, any feed back let me know, I'll try to fix it.

tillig commented on test/Autofac.Extras.DynamicProxy.Test/Autofac.Extras.DynamicProxy.Test.csproj:
```diff
@@ -44,6 +44,7 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.6.0" />
+    <PackageReference Include="Moq" Version="4.20.70" />
```
Can we use NSubstitute here instead of Moq? I am a fan of supporting open source, but some of the phone-home stuff that was implemented leaves Moq on vulnerability lists for some secure enterprises. In places where we _already had_ Moq, it should be pinned to the version just before that was implemented; for tests that are adding new mocking frameworks, we should go NSubstitute.

tillig commented on test/Autofac.Extras.DynamicProxy.Test/ServiceMiddlewareInterfaceInterceptorsFixture.cs:
```diff
@@ -0,0 +1,97 @@
+            .RegisterType<Interceptable>()
+            .InterceptedBy(typeof(StringMethodInterceptor))
+            .As<IPublicInterface>();
+        builder.EnableInterfaceInterceptors<IPublicInterface>();
```
I find this syntax to be confusing. We already registered a type to have interceptors, but then there's _another_ line to indicate interception.

I feel like it should be maybe more like...

```c#
builder.RegisterDecorator<T, U>().InterceptedBy(typeof(Z));
```

Something analogous to the current syntax.

idiotsky replied:
> I find this syntax to be confusing. We already registered a type to have interceptors, but then there's _another_ line to indicate interception.
> 
> I feel like it should be maybe more like...
> 
> ```cs
> builder.RegisterDecorator<T, U>().InterceptedBy(typeof(Z));
> ```
> 
> Something analogous to the current syntax.

If I change it to all in one line Is it ok? like this,

```
builder.RegisterType<Interceptable>().InterceptedBy(typeof(StringMethodInterceptor)).As<IPublicInterface>();
```

but I saw other file like InterfaceInterceptionWithPropertyInjectionFixture.cs

use 
```
 builder
            .RegisterType<InterceptableWithProperty>()
            .PropertiesAutowired()
            .EnableInterfaceInterceptors()
            .InterceptedBy(typeof(StringMethodInterceptor))
            .As<IPublicInterface>();
```
should we keep same style?

tillig replied:
The thing that makes the new syntax weird is that it's not connected to a RegisterType or RegisterDecorator. I think if it reads like a sentence - "register X as Y intercepted by Z" - that would be consistent and easier to understand.

tillig replied:
This is the bit currently pending. I'm not sure we want a syntax where it's some sort of builder level registration. I find the syntax to be challenging because there are now two ways to register things. It seems like if the interception needs to apply to a decorator, it should be attached to the decorator registration.

tillig replied:
Sorry, I've only had my phone, away for the holiday season, and I'm trying to answer quickly to help but also haven't been able to really look at anything on a big monitor and provide robust feedback.

In the test highlighted here, I see the new interception syntax like this:

```c#
// Register the decorator.
builder.RegisterDecorator<Decorator, IPublicInterface>();

// Register a type that needs to be intercepted.
builder
  .RegisterType<Interceptable>()
  .EnableInterfaceInterceptors()
  .InterceptedBy(typeof(StringMethodInterceptor))
  .As<IPublicInterface>();

// This is the confusing/inconsistent part - this standalone interception line.
builder.EnableInterfaceInterceptors<IPublicInterface>();
```

The thing that's inconsistent here is that, generally speaking, the syntax for registering things is all built up and attached to one registration.

```c#
// Start a registration
builder

  // Register a type `Interceptable`
  .RegisterType<Interceptable>()

  // I want to enable interface interception for this type
  .EnableInterfaceInterceptors()

  // Say what that type is intercepted by
  .InterceptedBy(typeof(StringMethodInterceptor))

  // Expose the type as `IPublicInterface`
  .As<IPublicInterface>();
```

It's a chained builder style syntax. I know that when I register the interceptor it will be intercepting things for `RegisterType`. I know the `As` applies to that registration.

The new line here:

```c#
builder.EnableInterfaceInterceptors<IPublicInterface>();
```

This is inconsistent. It reads like it's enabling some sort of _global_ interception, but if I read that in context of the other syntax...

- Why would I need `.EnableInterfaceInterceptors()` on the `RegisterType<T>` call if I'm supposed to use this new standalone `builder.EnableInterfaceInterceptors<T>()` syntax?
- What happens if I drop `.EnableInterfaceInterceptors()` on the `RegisterType<T>` call and _only_ use the new standalone syntax?
- I'm enabling interface interception, but by _what_? What's intercepting it? How do I control that?

I know that you can probably answer that here in comments, but that's not the point - the point is that someone later _reading the code_ or _trying to figure out how to make interception work_ is going to have these questions. The syntax shouldn't be confusing like that or cause folks to wonder what it means.

That's what I'm stuck on here.

If the goal of this PR is to allow interface interception on decorators, I would expect to see something more like:

```c#
[Fact]
public void InterceptsPublicInterfacesUseGenericMethod()
{
  ContainerBuilder builder = new();
  builder.RegisterType<StringMethodInterceptor>();

  // Same support for adding interception to a decorator
  // that we have for existing registrations. Familiar, consistent
  // syntax.
  builder
    .RegisterDecorator<Decorator, IPublicInterface>()
    .EnableInterfaceInterceptors()
    .InterceptedBy(typeof(StringMethodInterceptor);

  // Same syntax as before.
  builder
    .RegisterType<Interceptable>()
    .EnableInterfaceInterceptors()
    .InterceptedBy(typeof(StringMethodInterceptor))
    .As<IPublicInterface>();

  // ... and the rest of the test
}
```

Perhaps a good test to verify the syntax works is to see if we can get the _decorator_ to be intercepted by a _different interceptor_ than the thing being decorated.

```c#
[Fact]
public void DifferentInterceptorsForEachComponent()
{
  ContainerBuilder builder = new();
  builder.RegisterType<StringMethodInterceptor>();
  builder.RegisterType<CallCountInterceptor>();

  // Decorator gets intercepted by one thing
  builder
    .RegisterDecorator<Decorator, IPublicInterface>()
    .EnableInterfaceInterceptors()
    .InterceptedBy(typeof(CallCountInterceptor);

  // Base service gets intercepted by something else
  builder
    .RegisterType<Interceptable>()
    .EnableInterfaceInterceptors()
    .InterceptedBy(typeof(StringMethodInterceptor))
    .As<IPublicInterface>();

  // ... and the rest of the test
}
```

There are other features, too, that need to work the same if we add this.

- [Being able to add proxy generation options when enabling interception for the decorator.](https://github.com/autofac/Autofac.Extras.DynamicProxy/blob/2115f37e00614fa0507ad5dae984622558a6dc0f/test/Autofac.Extras.DynamicProxy.Test/InterceptorsChosenByMetadataWithOptionsFixture.cs#L19)
- [Being able to mark the decorator with an attribute to select the interceptor.](https://github.com/autofac/Autofac.Extras.DynamicProxy/blob/2115f37e00614fa0507ad5dae984622558a6dc0f/test/Autofac.Extras.DynamicProxy.Test/AttributedInterfaceInterceptionFixture.cs#L11)
- Possibly needing to differentiate between class and interface interceptors - can the class _being decorated_ have `EnableClassInterceptors` but the _decorator_ has `EnableInterfaceInterceptors`? What about the other way around?

That's the syntax and functionality that I'm stuck on, and it's why the standalone `builder.EnableInterfaceInterceptors<T>()` doesn't really work in the ecosystem.

tillig reviewed (commented):
I think the premise is good but I'm worried about the confusing syntax to enable this to happen. We might need to iterate over that a little to get it right.

Commit:
Switch from Moq to NSubstitute in unit tests.

Replaced Moq with NSubstitute in the test project to handle mocking.
Change RegisterType style in test.

Commit:
Refactor tests to remove NSubstitute and improve assertions

Removed the use of NSubstitute from the test project and simplified the testing structure. Adjusted assertions to account for the modified interceptor and decorator behavior, ensuring accurate validation of method outputs.
Fix ServiceMiddlewareInterfaceInterceptorsFixture test cases

idiotsky commented:
Although all checks have passed, I still have a few questions about the code.
ProxyHelpers line 61
```
if (!proxiedInterfaces.Any())
{
    return;
}
```
Actually, this code lacks code coverage. My question is regarding line 52: after calling EnsureInterfaceInterceptionApplies, can we confirm that proxiedInterfaces will never be empty? if yes, I think we can remove line 61 to 64.
ProxyHelpers line 52
```
EnsureInterfaceInterceptionApplies(ctx.Registration);

// The instance won't ever _practically_ be null by the time it gets here.
var proxiedInterfaces = ctx.Instance!
    .GetType()
    .GetInterfaces()
    .Where(ProxyUtil.IsAccessible)
    .ToArray();

if (!proxiedInterfaces.Any())
{
    return;
}
```

ProxyHelpers line 127, this code lacks code coverage too, I just worry about may be this code will never reach too.

```
if (!implTypeInfo.IsClass)
{
    return Enumerable.Empty<Service>();
}
```

tillig commented:
Re: Line 61 - the only difference I see is that when we check to see if interface interception applies, we're checking the service registration type, while when we check to see if there are any proxied interfaces, we're checking the actual instance type.

If memory serves, there was something with WCF proxy interception where we might end up with a service registration of, say, `IWcfService` but the object _instance_ was a [TransparentProxy](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.remoting.proxies.realproxy.gettransparentproxy?view=netframework-4.8.1) so the registration would say, "Yes, the interception applies!" but then an extension like WCF integration would have set up the instance and that instance _wouldn't_ support interception.

It's just my memory, though, I can't point you to a specific test or issue that indicated this. There was a lot with WCF proxies that wasn't awesome. I'm glad we don't have that so much in .NET Core but we still do support .NET Desktop.

If I had to guess, the way to test this would be to try creating a registration that has an event handler that replaces the instance with something like a TransparentProxy:

```c#
builder.RegisterType<MyType>()
       .As<IMyInterface>()
       .OnActivating(e => e.ReplaceInstance(new TransparentProxy()));
```

Something like that - the idea being the instance doesn't actually implement the interfaces the way we expect. I don't remember the exact syntax off the top of my head.

I won't hold up the PR on that. Some of these runtime things can be hard to replicate.

Re: Line 127 - I think you're right, the implementation type will only ever be a class so that line should never practically be hit. Nice catch!

Minor point - we probably want to make that `ProxyHelpers` class internal. I'm not sure we want to support folks outside the library calling it.

Commit:
Make `ProxyHelpers` internal and improve code clarity.

Refactored `ProxyHelpers` to be `internal` for better encapsulation. Fixed a typo in exception documentation and removed unreachable code in `GetInterceptorServicesFromAttributes()`. Added `InternalsVisibleTo` for unit testing support.

idiotsky commented:
I already change ProxyHelpers to internal and remove line 127 and related code.
And also, I try 
Add this to InterfaceInterceptorsFixture
'''
[Fact]
public void InterceptsPublicInterfacesWithReplacedInstance()
{
    var builder = new ContainerBuilder();
    builder.RegisterType<StringMethodInterceptor>();
    builder
        .RegisterType<Interceptable>()
        .InterceptedBy(typeof(StringMethodInterceptor))
        .EnableInterfaceInterceptors()
        .As<IPublicInterface>()
        .OnActivating(e => e.ReplaceInstance(new TransparentProxy()));
    var container = builder.Build();
    var obj = container.Resolve<IPublicInterface>();
    Assert.Equal("intercepted-PublicMethod", obj.PublicMethod());
}
public class TransparentProxy
{
}
'''
while debug the  ctx.Instance( in ProxyHelpers line 55) still Interceptable not TransparentProxy when call container.Resolve<IPublicInterface>(); it throws excpetion as following
Autofac.Core.DependencyResolutionException: An exception was thrown while activating Autofac.Extras.DynamicProxy.Test.InterfaceInterceptorsFixture+Interceptable.
 ---> System.InvalidCastException: Unable to cast object of type 'Castle.Proxies.IPublicInterfaceProxy' to type 'Interceptable'.

So may be this is not the case.

tillig commented:
Well, like I said, I can't remember the exact syntax for `TransparentProxy` off the top of my head. I know when this is used in context of WCF, it _somehow_ creates an object that implements the WCF service interface. It's not a super common use case, but it does happen.

Don't worry about coverage on that one line. I'm a little bit afraid of removing it, though, since, again, I have a vague memory from many years ago where we had to add that for _something_.

idiotsky commented:
> Well, like I said, I can't remember the exact syntax for `TransparentProxy` off the top of my head. I know when this is used in context of WCF, it _somehow_ creates an object that implements the WCF service interface. It's not a super common use case, but it does happen.
> 
> Don't worry about coverage on that one line. I'm a little bit afraid of removing it, though, since, again, I have a vague memory from many years ago where we had to add that for _something_.

ok, nice, can you approve the pull request or any comment on it?

tillig commented:
The comment on the syntax is what's holding this up.

idiotsky commented:
> The comment on the syntax is what's holding this up.
if you mean this
```
builder.RegisterDecorator<T, U>().InterceptedBy(typeof(Z));
```
the RegisterDecorator mehod is from autofac main project and currently no return value. so it can not be chained.

tillig commented:
Thank you for this, @idiotsky — and for your patience through the back-and-forth. This PR did the important work of proving the concept out and surfacing exactly where the friction is, and that directly shaped the plan we've now landed on.

After digging into the internals, here's where we netted out:

- **The feature is worth doing.** Being able to intercept a decorator is a legitimate, useful capability.
- **The standalone `builder.EnableInterfaceInterceptors<TService>()` syntax isn't the direction we want to take**, for the reasons discussed in the review — it's disconnected from the decorator registration, so it's ambiguous what's being intercepted and by what. We want the familiar chained style: `builder.RegisterDecorator<Decorator, IService>().EnableInterfaceInterceptors().InterceptedBy(typeof(X))`.
- **The real blocker is in core Autofac, not here.** `RegisterDecorator<TDecorator, TService>()` returns `void` and snapshots the decorator registration's pipeline internally, so there's nothing to attach interception to. To do this cleanly, core needs to return a (narrow, decorator-specific) configurable builder — which is a binary-breaking change that has to be timed for a major core release.

I've written all of this up so the effort isn't lost:

- Downstream feature: autofac/Autofac.Extras.DynamicProxy#66
- Core change + plan: autofac/Autofac#1416 (with a detailed breakdown in the latest comment)

A couple of pieces of this PR are genuinely reusable and called out in #66 — notably extracting the proxy logic into a shared `ProxyHelpers` — and you're very welcome to take a run at any of the follow-up work once the core change is in place.

Given the above, I'm going to close this PR in favor of those issues, which capture the agreed approach and dependencies. Closing isn't a reflection on the effort here — it was the catalyst for the whole plan. Thanks again. 🙏
